util/progress/progressui/display.go

```diff
@@ -71,53 +83,152 @@ func DisplaySolveStatus(ctx context.Context, c console.Console, w io.Writer, ch
 
 	var done bool
 	ticker := time.NewTicker(tickerTimeout)
-	// implemented as closure because "ticker" can change
-	defer func() {
-		ticker.Stop()
-	}()
+	defer ticker.Stop()
 
 	displayLimiter := rate.NewLimiter(rate.Every(displayTimeout), 1)
 
-	var height int
-	width, _ := disp.getSize()
 	for {
 		select {
 		case <-ctx.Done():
 			return nil, ctx.Err()
 		case <-ticker.C:
 		case ss, ok := <-ch:
 			if ok {
-				t.update(ss, width)
+				d.update(ss)
 			} else {
 				done = true
 			}
 		}
 
-		if modeConsole {
-			width, height = disp.getSize()
-			if done {
-				disp.print(t.displayInfo(), width, height, true)
-				t.printErrorLogs(c)
-				return t.warnings(), nil
-			} else if displayLimiter.Allow() {
-				ticker.Stop()
-				ticker = time.NewTicker(tickerTimeout)
-				disp.print(t.displayInfo(), width, height, false)
-			}
-		} else {
-			if done || displayLimiter.Allow() {
-				printer.print(t)
-				if done {
-					t.printErrorLogs(w)
-					return t.warnings(), nil
-				}
-				ticker.Stop()
-				ticker = time.NewTicker(tickerTimeout)
-			}
+		if done {
+			d.display(true)
+			return d.warnings(), nil
+		} else if displayLimiter.Allow() {
+			d.display(false)
+			ticker.Reset(tickerTimeout)
 		}
 	}
 }
 
+type consoleDisplay struct {
+	t             *trace
+	disp          *display
+	width, height int
+}
+
+func NewConsoleDisplay(c console.Console) SolveStatusDisplay {
+	return &consoleDisplay{
+		t:    newTrace(c, true),
+		disp: &display{c: c},
+	}
+}
+
+func (d *consoleDisplay) init(dsso *displaySolveStatusOpts) {
+	d.disp.phase = dsso.phase
+	d.disp.desc = dsso.consoleDesc
+	if d.disp.phase == "" {
+		d.disp.phase = "Building"
+	}
+	d.width, d.height = d.disp.getSize()
+}
+
+func (d *consoleDisplay) update(ss *client.SolveStatus) {
+	d.t.update(ss, d.width)
+}
+
+func (d *consoleDisplay) display(done bool) {
+	d.width, d.height = d.disp.getSize()
+	d.disp.print(d.t.displayInfo(), d.width, d.height, done)
+	if done {
+		d.t.printErrorLogs(d.t.w)
+	}
+}
+
+func (d *consoleDisplay) warnings() []client.VertexWarning {
+	return d.t.warnings()
+}
+
+type textDisplay struct {
+	t       *trace
+	printer *textMux
+}
+
+func newTextDisplay(w io.Writer, printer vertexPrinter) SolveStatusDisplay {
+	return &textDisplay{
+		t:       newTrace(w, false),
+		printer: newTextMux(w, printer),
+	}
+}
+
+func newJSONDisplay(w io.Writer) SolveStatusDisplay {
+	return newTextDisplay(w, &jsonPrinter{})
```

Comment:
I need to check this more but it looks to me that this `newTextDisplay` means that `newTextMux` is used for JSON as well. TextMux is the thing that tries print multiple parallel steps together in one text stream by making one of them active and then with control variables like `antiFlicker/maxDelay` etc controls how often the active item switches and when partial data is printed to avoid too much buffering. For JSON, I don't think this is ideal and we should print new JSON progress objects once the vertex goes to `Complete` state.